### PR TITLE
pr: harden post-processing against dictionary, list, and four-digit number formatting false positives

### PR DESCRIPTION
**Pull request by macmixing**
## Summary
This branch tightens several deterministic post-processing paths in `KeyVoxCore` to reduce false positives without weakening the valid formatting/correction cases we still want to preserve.

It covers three main areas:
1. Dictionary matching for stylized split-join brand corrections
2. List detection and trailing list-item formatting
3. Four-digit quantity grouping for sub-10,000 numerals

### Dictionary matcher
- Added a tail-similarity guard to the anchored stylized split-join scoring path in `DictionaryMatcher+SplitJoinScoring.swift`.
- This keeps valid Whisper-style split brand corrections working, but blocks false positives where ordinary prose gets collapsed into a stylized dictionary term.
- Added a regression for the `"key value"` style false positive so generic two-word prose does not get rewritten into a branded token.

### List detection and formatting
- Hardened ambiguous undelimited marker detection in `ListPatternDetector.swift` so short low-content spans between spoken numbers do not get promoted into lists.
- This prevents prose like `"two or three points taller"` from being misread as numbered list cadence.
- Improved last-item trailing split behavior so short nominal final items like `"Bananas"` can cleanly separate from trailing commentary such as `", and then I need..."`.
- Re-enabled terminal punctuation for longer list items while still stripping terminal punctuation from very short one- and two-word items, which keeps short nominal items clean without flattening sentence-like list entries.
- Added detector and rendering regressions for:
  - uncertain numeric-range prose that should stay prose
  - explicit-number and spoken-marker lists with trailing commentary after the last item
  - mixed short/long list items where longer entries should retain terminal punctuation

### Four-digit quantity grouping
- Added a dedicated `ThousandsGroupingNormalizer` and wired it into `TranscriptionPostProcessor` after website normalization and before final whitespace/casing cleanup.
- The new pass formats quantity-like four-digit numerals such as `5600` -> `5,600`.
- Year-like references remain ungrouped, including conversational year phrasing and sentence-final uncertainty/confirmation shapes.
- Protected ranges prevent grouping inside date/version/phone/compact hyphenated numeric formats.
- Added regressions covering:
  - standalone quantity sequences
  - mixed quantity/year sentences
  - partitive quantity phrases like `2000 of them`
  - confirmation phrases like `2001, right?`
  - uncertain year references like `maybe 1993`

### Documentation
- Updated `Docs/CODEMAP.md` to include the new numeric grouping pass.
- Refined `Docs/ENGINEERING.md` so normalization is documented as stable ordered stages instead of an ever-growing list of micro-passes.

## Why
These changes reduce brittle output transformations in three places users actually notice:
- branded dictionary corrections should not rewrite ordinary prose
- prose with incidental spoken numbers should not become lists
- quantities below 10,000 should read naturally with grouping, while years should remain untouched

## Testing
- Added regression coverage across dictionary, list detector, list formatting, and transcription post-processing tests
- Verified with:
  `swift test --package-path Packages/KeyVoxCore --scratch-path /tmp/keyvoxcore-tests`

## Summary by CodeRabbit

* **New Features**
  * Automatic thousands grouping for four-digit+ numeric sequences in transcriptions, while preserving years, dates, versions, and phone-like spans.

* **Bug Fixes**
  * Improved list detection and punctuation handling for varied item lengths and comma-boundary cases.
  * Tighter matching heuristics for certain stylized anchored token cases.
  * Ensured local hyphenated phone numbers are preserved from normalization.

* **Tests**
  * Added coverage for numeric grouping, list formatting, matcher behavior, and phone-number protections.

* **Documentation**
  * Updated processing pipeline docs to reflect staged normalization and the new grouping stage.

### CODE REVIEW

**Comment by macmixing** [comment]
@codex review

File: Packages/KeyVoxCore/Sources/KeyVoxCore/Lists/ListPatternDetector.swift [hunk 1]
```diff
@@ -113,23 +122,63 @@ public struct ListPatternDetector {
...
+
+        return withoutStructuralDelimiter
             .replacingOccurrences(
-                of: #"[\"'”’\)\]\}]*[.,;:!?…]+[\"'”’\)\]\}]*$"#,
+                of: #"[\"'”’\)\]\}]*[.!?…]+[\"'”’\)\]\}]*$"#,
                 with: "",
                 options: .regularExpression
             )
             .trimmingCharacters(in: .whitespacesAndNewlines)
     }
```
[the remark this answers is not in the document]
Comment (macmixing) [reply Packages/KeyVoxCore/Sources/KeyVoxCore/Lists/ListPatternDetector.swift:145]: whisper local base model does not output this and this is a non issue @coderabbitai

File: Packages/KeyVoxCore/Sources/KeyVoxCore/Normalization/ThousandsGroupingNormalizer.swift [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+        if next == nil,
+           let previousTag = previous?.tag,
+           [.noun, .determiner, .preposition].contains(previousTag) {
+            return true
+        }
+
+        if previous?.tag == .determiner, next?.tag == .noun {
+            return true
+        }
```
[the remark this answers is not in the document]
Comment (macmixing) [reply Packages/KeyVoxCore/Sources/KeyVoxCore/Normalization/ThousandsGroupingNormalizer.swift]: @coderabbitai hard-coded temporal nouns are against the codebase rules.

[the remark this answers is not in the document]
Comment (macmixing) [reply Packages/KeyVoxCore/Sources/KeyVoxCore/Normalization/ThousandsGroupingNormalizer.swift]: @coderabbitai we must resolve this without them, what are your suggestions?
